[NFC] Remove Type::getInt8PtrTy

pmatos opened the pull request:
Replace this with PointerType::get().
Also adds AS = 0 as an optional argument to get() methods.

Followup to the opaque pointer transition. Fixes an in-code TODO item.

nikic commented:
> Also adds AS = 0 as an optional argument to get() methods.

I believe that was intentionally omitted to make sure address spaces aren't forgotten (getUnqual is used for the case where they aren't relevant). cc @arsenm 

Not sure whether this consideration is still relevant with opaque pointers, where we no longer create random pointer types for bitcasts all over the place. It may be better to accept the 0 default for simplicity.

arsenm commented:
I still think all the defaulted address space parameters should be purged and it only invites bugs. It's a regression to introduce a new defaulted argument.

pmatos commented:
> I still think all the defaulted address space parameters should be purged and it only invites bugs. It's a regression to introduce a new defaulted argument.

Understandable, although always writing `PointerType::get(Ctx, 0);` really looks weird given most of the time the second parameter is 0. So it just feels like it's a special constant that will suddenly be sprinkled all over the code.

mtrofin commented:
Nit: can you add in the description that this is a follow-up from the opaque pointer transition (having an explicit motivation helps read a patch). Thanks!

pmatos commented:
> Nit: can you add in the description that this is a follow-up from the opaque pointer transition (having an explicit motivation helps read a patch). Thanks!

you're right. done.

JOE1994 commented:
I have a similar concern with @arsenm about making the default address space 0.
Can't we just use the already existing `PointerType::getUnqual()`, instead of introducing the default address space 0?

pmatos commented:
> I have a similar concern with @arsenm about making the default address space 0. Can't we just use the already existing `PointerType::getUnqual()`, instead of introducing the default address space 0?

True, we could indeed just call `PointerType::getUnqual()` instead of just `PointerType::get()`.

pmatos committed:
[NFC] Remove Type::getInt8PtrTy

Replace this with PointerType::getUnqual().
Followup to the opaque pointer transition. Fixes an in-code TODO item.

pmatos commented:
I pushed forced a change using PointerType::getUnqual. Would this be more acceptable? It certainly looks cleaner to my eyes than having `PointerType::get(Ctx, 0)` everywhere.

JOE1994 commented on llvm/lib/Transforms/Scalar/RewriteStatepointsForGC.cpp:
```diff
@@ -1514,7 +1514,7 @@ static void CreateGCRelocates(ArrayRef<Value *> LiveVariables,
     assert(isHandledGCPointerType(Ty, GC));
     auto AS = Ty->getScalarType()->getPointerAddressSpace();
-    Type *NewTy = Type::getInt8PtrTy(M->getContext(), AS);
+    Type *NewTy = PointerType::getUnqual(M->getContext(), AS);
```
Thank you for the update 👍 

The following update is needed to avoid build failure.
```suggestion
    Type *NewTy = PointerType::get(M->getContext(), AS);
```

pmatos replied:
Right - thanks for pointing that out.

nikic commented:
You also need to replace uses in polly.

JOE1994 commented:
https://buildkite.com/llvm-project/github-pull-requests/builds/12872#018ba3cf-2050-4af1-b46a-55b5482323fc

JOE1994 reviewed (changes requested):
Please update existing uses in `polly`, and rebase onto latest `main`.

JOE1994 reviewed (approved):
Thank you

dtcxzyw commented:
Please also update existing uses in `lldb`.
Buildbot failure: https://lab.llvm.org/buildbot/#/builders/68/builds/63101

pmatos commented:
> Please also update existing uses in `lldb`. Buildbot failure: https://lab.llvm.org/buildbot/#/builders/68/builds/63101

Please see https://github.com/llvm/llvm-project/pull/71561
